cas-compute/src/symbolic/expr/mod.rs

```diff
@@ -441,50 +441,52 @@ impl PartialEq for Expr {
+/// Implements the [`Hash`] trait for the [`Expr`] enum, which represents
 /// different types of mathematical expressions. This implementation is necessary
-/// because [`PartialEq`] is also manually implemented, and the automatically
-/// derived [`Hash`] implementation would not guarantee consistency with [`PartialEq`].
-impl std::hash::Hash for Expr {
-    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
-        // Hash the discriminant to differentiate between enum variants.
+/// because [`PartialEq`] is also manually implemented. The derived implementation
+/// of [`Hash`] would not guarantee consistency with [`PartialEq`], especially for 
+/// commutative operations like addition and multiplication.
+impl Hash for Expr {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        // Hash the discriminant to differentiate between enum variants
         let self_discr = std::mem::discriminant(self);
-        std::hash::Hash::hash(&self_discr, state);
+        Hash::hash(&self_discr, state);
 
-        // Hash the data associated with each variant.
+        // Hash the data associated with each variant
         match self {
-            Expr::Primary(val) => std::hash::Hash::hash(val, state),
+            Expr::Primary(val) => val.hash(state),
             Expr::Add(val) | Expr::Mul(val) => {
                 // For Add and Mul, the order of the elements should not affect the hash.
-                // Therefore, we'd have to sort the individual elemets hashes before
-                // hashing them.
-
-                let mut hashes: Vec<u64> = val
-                    .iter()
-                    .map(|expr| {
-                        let mut hasher = DefaultHasher::new();
-
-                        std::hash::Hash::hash(expr, &mut hasher);
-
-                        hasher.finish()
-                    })
-                    .collect();
-
-                // To satisfy order invariance.
-                hashes.sort_unstable();
-
-                for hash in hashes {
-                    std::hash::Hash::hash(&hash, state);
+                // To handle commutativity, we'll hash the elements using both summation
+                // and multiplication of individual hashes, which will remain invariant
+                // under different orders of the same elements.
+
+                let mut sum_hash = 0u64;
+                let mut prod_hash = 1u64;
+
+                // Use a single hasher for all expressions to avoid repeatedly creating new ones.
+                for expr in val {
+                    let mut hasher = DefaultHasher::new();
+                    expr.hash(&mut hasher);
+                    let expr_hash = hasher.finish();
+
+                    sum_hash = sum_hash.wrapping_add(expr_hash);
+                    // Avoid multiplication by `0` by adding `1` to each hash
+                    prod_hash = prod_hash.wrapping_mul(expr_hash.wrapping_add(1));
```

Comment:
It's inevitable that a value will hash to 0, so it's not that particularly important to check for it. The line `expr_hash.wrapping_add(1)` can result in 0 anyway, if `expr_hash` is equal to `u64::MAX`. Instead, the goal should be to spread the resulting hashes out as much as possible, even if a hash value of 0 appears. See my below comment for a suggestion.

Reply:
Do you think simply summing the individual hashes would suffice? i.e be better than XOR-ing?

The hash function should ideally distribute values uniformly across inputs. So, `hash(a) + hash(b)` seems to be valid approach to me. As it wouldn't just remove the effect of even values. e.g hash collision example with `XOR` implementation: `hash(Sum(a, x, a)) == hash(Sum(b, b, x))` so, in this if internal hashing algorithm is implemented using `XOR` there for sure would be a collision.

Since I've combined the `Add` and `Mul` variants (as they has similar signatures) under a single case, the hash value would indeed be identical for both of these unless you include additional information like `std::mem::discriminant(self)` to distinguish them. Yes, the discriminant is already being used to differentiate and diversify/spread the hash, but separating `Add` and `Mul` with distinct logic for hashing could improve the hash function's distribution (or maybe I am over analyzing).

Also, we have to take order in-variance into account for commutative operations such as above.

Reply:
Summing the hashes would certainly suffice. Our approach so far relies on using `std::hash::DefaultHasher`, which currently is SipHash 1-3, to hash each individual expression within the collection. Assuming that those hashes are randomly and uniformly distributed across `u64`, summing them up (while allowing them to wrap) will result in a new hash that is also randomly and uniformly distributed. As we've also established, summing is commutative, which is exactly what we need for our problem.

XORing hashes on its own would be more problematic since expressions that occur an even number of times would cancel each other out, as you described.

Hashing the discriminant should be enough to ensure that each enum variant will surely hash differently. If we _didn't_ trust SipHash 1-3, only _then_ would it probably be necessary to modify the hash function for `Add` and `Mul`!

Also, what I meant by my above comment was that adding special logic to avoid multiplication by zero is not necessary if we rely on the idea of summing randomly and uniformly distributed hashes. In fact, the approach of multiplying by `expr_hash.wrapping_add(1)` can still result in zero is `expr_hash` is `u64::MAX`, which would cause the whole hash to degenerate into zero. This is, of course, practically impossible to achieve, but still worth mentioning. A more robust approach would be to simply sum the hashes.

Reply:
Considering all this advice I have simplified the function further and all tests are passing well. Still some input would be much appreciated.